File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/MicQueueView/PanoMicQueueView.m
//
//  PanoMicQueueView.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoMicQueueView.h"
#import "PanoMicInfo.h"
#import "PanoMicCollectionCell.h"

CGFloat InteritemSpacing  = 25;
CGFloat LineSpacing = 30;
CGFloat MarginSpacing = 30;

@interface PanoMicQueueView () <UICollectionViewDataSource,
                                UICollectionViewDelegateFlowLayout>
@end

@implementation PanoMicQueueView {

    UICollectionView *collectionView;

    UICollectionViewFlowLayout *layout;
}



- (void)initViews {
    
    layout = [[UICollectionViewFlowLayout alloc] init];
    layout.minimumInteritemSpacing = 0.1;
    collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0, 0, 0) collectionViewLayout:layout];
    collectionView.backgroundColor = [UIColor clearColor];
    collectionView.showsVerticalScrollIndicator = false;
    collectionView.showsHorizontalScrollIndicator = false;
    collectionView.dataSource = self;
    collectionView.delegate = self;
    collectionView.bounces = false;
    collectionView.clipsToBounds = false;
    [collectionView registerClass:[PanoMicCollectionCell class] forCellWithReuseIdentifier:@"cellID"];
    [self addSubview:collectionView];
}

- (void)initConstraints {
    [collectionView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(0).insets(UIEdgeInsetsMake(0, MarginSpacing, 0, MarginSpacing));
    }];
}

- (void)reloadData {
    [collectionView reloadData];
}

- (CGSize)cellSize {
    CGFloat width = (PanoAppWidth - 3 * LineSpacing - 2 * MarginSpacing) / 4;
    if (width > 75) {
        width = 75;
    }
    CGFloat height = width + 30;
    return CGSizeMake(width, height);
}

#pragma mark -- UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return data.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    PanoMicCollectionCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"cellID" forIndexPath:indexPath];
    cell.delegate = _delegate;
    [cell updateMicInfo: data[indexPath.row]];
    return cell;
}

#pragma mark -- UICollectionViewDelegateFlowLayout
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return [self cellSize];
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return LineSpacing;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return MarginSpacing;
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/MicQueueView/PanoMicQueueView.swift
//
//  PanoMicQueueView.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/10/29.
//

import UIKit


class PanoMicQueueView: PanoBaseView, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    
    let InteritemSpacing : CGFloat = 25

    let LineSpacing : CGFloat = 30
    
    let MarginSpacing : CGFloat = 30
    
    var data = [PanoMicInfo]();
    
    var collectionView: UICollectionView!
    
    var layout: UICollectionViewFlowLayout!
    
    weak var delegate: PanoMicCollectionCellDelegate?
    
    override func initViews() {
        
        layout = UICollectionViewFlowLayout()
        layout.minimumInteritemSpacing = 0.1
        
        collectionView = UICollectionView(frame: CGRect(x: 0, y: 0, width: 0, height: 0), collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.clear
        collectionView.showsVerticalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.bounces = false
        collectionView.clipsToBounds = false
        collectionView.register(PanoMicCollectionCell.self, forCellWithReuseIdentifier: "cellID")
        self.addSubview(collectionView)
    }
    
    override func initConstraints() {
        collectionView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview().inset(UIEdgeInsets(top: 0, left: MarginSpacing, bottom: 0, right: MarginSpacing))
        }
    }
    
    func reloadData()  {
        collectionView.reloadData()
    }
    
    func cellSize() -> CGSize {
        var width = (PanoAppWidth - 3 * LineSpacing - 2 * MarginSpacing) / 4
        if width > 75 {
            width = 75
        }
        let height = width + 30;
        return CGSize(width: width, height: height)
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        let count = data.count
        return count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cellID", for: indexPath) as! PanoMicCollectionCell
        cell.delegate = delegate
        let info = data[indexPath.row]
        cell.updateMicInfo(info: info)
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return cellSize()
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return LineSpacing
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return MarginSpacing
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        
    }
    
}


